### PR TITLE
Change non-git shell script files to node files

### PR DESCRIPTION
**Pull request by mattferrin**
I think this is a good step in the right direction. I tested it on my Windows an Mac machines.

The most controversial change, I switched the example to run on port 8081 so that it is different from the port used in the main non-example.

### CODE REVIEW

File: example/server.js [hunk 1]
```diff
@@ -15,4 +15,4 @@ const app = express();
 app.use(express.static(__dirname));
 app.use('/graphql', graphqlHTTP(() => ({ schema })));
 
-app.listen(8080, () => console.log('Started on http://localhost:8080/'));
+app.listen(8081, () => console.log('Started on http://localhost:8081/'));
```
Comment (wincent) [inline example/server.js]: As this is unrelated, should probably be pulled out into a separate commit.

In any case, I actually think we should use a port number of 0 here, which will cause the OS to assign a random port number. I've just sent a PR for that: https://github.com/graphql/graphiql/pull/321

**Comment by wincent** [comment]
Thanks for taking the time to get things working on Windows, @mattferrin! I might add a comment or two in line, but here are some high level remarks to start with:

This is a good start. I think, however, that we should address my suggestions from https://github.com/graphql/graphiql/issues/318 before merging this. It seems quite wrong, for example, to be shelling out to the `echo` executable from node when we could instead just print to standard out with `process.stdout.write`. Likewise, there are "native" Node equivalents for much of the shell stuff we are doing in here (the `fs.*` methods I suggested in the other issue) and existing, mature abstractions to replace forking out to platform-specific commands  (for example, we could use `rimraf` instead of `rm -rf`/`rd`).

This isn't so much a "port to node" so much as a "writing bash in node", which is pretty painful to write, and even read (due to all the escaping). I'm happy to help out with doing an a node-centric rewrite if you're not up for it.

Thanks once again for your time on this!

File: resources/prepublish.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+if (process.env.npm_config_argv) {
+  let npmConfigArgv = JSON.parse(process.env.npm_config_argv);
+  // # Because of a long-running npm issue (https://github.com/npm/npm/issues/3059)
+  // # prepublish runs after `npm install` and `npm pack`.
+  // # In order to only run prepublish before `npm publish`, we have to check argv.
+  if (npmConfigArgv.original[0] === 'publish') {
+    // # Publishing to NPM is currently supported by Travis CI, which ensures that all
+    // # tests pass first and the deployed module contains the correct file structure.
+    // # In order to prevent inadvertently circumventing this, we ensure that a CI
+    // # environment exists before continuing.
```
Comment (wincent) [inline resources/prepublish.js]: Can you remove this Bash comment syntax from inside your JavaScript comments?

File: example/server.js (hunk 1 above)
Comment (mattferrin) [inline example/server.js]: Undone.

**Comment by mattferrin** [comment]
I've removed some bash style comments because I could swear I was asked to, though I can no longer seem to find that request to comment that it's done. 

As for the port to Node file system functions, I do not have the time for it myself. My goal is to keep my code as close as possible to the existing and working solution and not break things.

**Comment by wincent** [comment]
> I've removed some bash style comments because I could swear I was asked to, though I can no longer seem to find that request to comment that it's done.

[Hidden away here](https://github.com/graphql/graphiql/pull/320#discussion_r100392463).

Thanks for the updates!

File: resources/prepublish.js (hunk 2 above)
Comment (mattferrin) [inline resources/prepublish.js]: Gotcha. Done :)

**Comment by mattferrin** [comment]
FYI, I'm doing a little more work on this even though I said I didn't have time, and I don't! :)

**Comment by mattferrin** [comment]
@wincent Your suggestion 

> Well, it is literally just concatenating the files, so could get the listing with fs.readdirSync, read the files with fs.readFileSync, concat in JS, then fs.writeFileSync the result.

wasn't implemented because I don't know how to do it while file globbing.

**Comment by mattferrin** [comment]
@wincent Apologies. I'm blind. You explicitly state I can use `readdirSync`. I'll do that. Thanks.

**Comment by mattferrin** [comment]
I think I've completed all the suggestions. If you would like I can soft reset all my commits and make it one commit, after we're done collaborating on this of course.

**Comment by wincent** [comment]
> FYI, I'm doing a little more work on this even though I said I didn't have time, and I don't! :)

I know the feeling well.

> I think I've completed all the suggestions. If you would like I can soft reset all my commits and make it one commit, after we're done collaborating on this of course.

That won't be necessary. GitHub has a shiny "Squash and merge" button. I have reviewed the new changes yet, but will do so shortly. Thanks for taking this extra stab at it!

File: resources/build.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+var isWindows = /^win/.test(process.platform)
+
+var out = function (command) {
+  execSync(command, {stdio:'inherit'});
+};
+
+console.log('"Running node ./resources/build.js"');
+
+if (!isWindows) {
+  out('set -e'); // exit on console error
```
Comment (wincent) [inline resources/build.js]: I don't think this will do anything, will it?

`out` will fork a child process that lives for the duration of the `set` command, then exit, so the attempt to change the setting will have no subsequent effect.

I think we can delete this.

File: resources/build.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+if (!fs.existsSync('node_modules/.bin')) {
+  console.log('"Be sure to run \`npm install\` before building GraphiQL."');
+}
+
+var distDir = './dist';
+rimraf.sync(distDir);
+fs.mkdirSync(distDir);
+
+out('babel src --ignore __tests__ --out-dir dist/');
+console.log('"Bundling graphiql.js..."');
```
Comment (wincent) [inline resources/build.js]: Why the double quotes here?

File: resources/build.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+out('browserify -g browserify-shim -s GraphiQL dist/index.js > graphiql.js');
+console.log('"Bundling graphiql.min.js..."');
```
Comment (wincent) [inline resources/build.js]: And here (etc; I won't bother commenting on the other places).

File: resources/build.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+if (isWindows) {
+  out('browserify -g browserify-shim -g uglifyify -s GraphiQL dist/index.js | uglifyjs -c --screw-ie8 > graphiql.min.js');
+} else {
+  out('browserify -g browserify-shim -g uglifyify -s GraphiQL dist/index.js 2> /dev/null | uglifyjs -c --screw-ie8 > graphiql.min.js 2> /dev/null');
```
Comment (wincent) [inline resources/build.js]: I think we can do this without piping to the shell but using real pipes in a platform-agnostic way, so we won't need the `isWindows` check here.

File: resources/build.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
+var execSync = require('child_process').execSync;
```
Comment (wincent) [inline resources/build.js]: Just a thought: you could use ES6 features in this file if you executed it with `babel-node` instead of `node`.

Comment (mattferrin) [inline resources/build.js]: It's a good idea. I'll definitely do that on personal projects moving forward, but am unsure it's best here.

File: resources/build.js (hunk 4 above)
Comment (mattferrin) [inline resources/build.js]: They are accidental remnants from past echo commands. I can remove them, but noticed they seemed to help contrast between other console writes and made these easier to read even if the quotes lacked meaning. (There are JS filenames logged to console during the build without quotes.)

File: resources/build.js (hunk 5 above)
Comment (mattferrin) [inline resources/build.js]: Confirm you'd like me to remove them and the double quotes will be gone :)

File: resources/build.js (hunk 6 above)
Comment (mattferrin) [inline resources/build.js]: You're right. It might take a while for me to get around to doing this.

File: resources/build.js (hunk 7 above)
Comment (mattferrin) [inline resources/build.js]: Say to do it and I'm game though.

File: resources/build.js (hunk 3 above)
Comment (mattferrin) [inline resources/build.js]: I remember it halting program execution upon an error, but I need to find time to test and prove my memory is either correct or incorrect.

Comment (wincent) [inline resources/build.js]: Sounds good. I would be extremely surprised if it has any effect.

File: resources/build.js (hunk 5 above)
Comment (wincent) [inline resources/build.js]: Yeah, I think you should.

File: resources/build.js (hunk 7 above)
Comment (wincent) [inline resources/build.js]: We have Babel in the project precisely so that we can use ES6 features, so doing so would be consistent with the other usage in the repo.

Comment (asiandrummer) [inline resources/build.js]: Let's do `babel-node` for consistency, as @wincent suggested!

**Comment by asiandrummer** [comment]
I don't feel qualified enough to review this as I have no prior experience working with Windows env - I'll trust @wincent and @mattferrin's judgements on merging this one.

**Comment by mattferrin** [comment]
@asiandrummer It looks like I have mostly minor changes to make, but am unsure how to tackle the platform agnostic piping thing in Node off hand. 

So basically, I just haven't prioritized my time to include finishing this as of yet.

**Comment by holm0563** [comment]
Not sure about mac but i can verify that this is the only way i can build on a windows machine.

**Comment by AGS-** [comment]
Hey @mattferrin, sorry we let this PR go a little stale. Would you be willing to rebase your changes onto master? I do have both a Windows and OSX machine I can test the PR on. Let me know if you're still willing to work on this. Thanks for contributing!

**Comment by mattferrin** [comment]
Hi @AGS-, my Windows machine was a prior company's computer. While it seems strange to actually say it, I don't have access to one. I know my code was functional, just slightly hacky. Being from a busy small team, didn't make sense to convert it all to pure Node. Sorry, but without a Windows machine and little incentive, I think I'm a dead end on this thread 🙂 Thanks for the renewed interest though.

**Comment by AGS-** [comment]
Oh, that's unfortunate. Closing for now and hoping someone picks it up later on. Thanks for having taken the time to contribute!
